### PR TITLE
Add destroy method to NcclCommunicator

### PR DESCRIPTION
**Pull request by kuenishi**
TL; DR: this pull request is for fault tolerance in ChainerMN.

To continue distributed training in ChainerMN, the training cluster must endure process failure and also the communicator must be reconstructed. The simplest way to enable reconstruction is to add explicit `destroy()` method, rather than reconstruction method. This change enables reconstruction like this:

```python
  comm = nccl.NcclCommunicator(8, comm_id, 3)
  ...(node failure detected)...
  comm.destroy()
  new_rank = renew_rank()
  comm = nccl.NcclCommunicator(7, comm_id, new_rank)
  ...(resume computation)...
```

See [another example](https://gist.github.com/kuenishi/370df96974051b2dc6159366a8f62d7d) for working code. Maybe I could add null-checking code for all other methods than `destroy()`, but don't want to mess up current code more than this before basic discussion.

### CODE REVIEW

File: cupy/cuda/nccl.pyx [hunk 1]
```diff
@@ -127,8 +127,12 @@ cdef class NcclCommunicator:
         check_status(status)
 
     def __dealloc__(self):
+        self.destroy()
+
+    def destroy(self):
```
Comment (okuta) [inline cupy/cuda/nccl.pyx]: Please use `cpdef`

**Comment by okuta** [comment]
LGTM!
